common/ringpop/ringpop.go

```diff
@@ -34,131 +35,186 @@ import (
+					} else {
+						dialer := tls.Dialer{Config: clientTLSConfig}
+						if tChannel, err := tchannel.NewChannel(ringpopServiceName, &tchannel.ChannelOptions{Dialer: dialer.DialContext}); err != nil {
+							factory.logger.Fatal("Failed to create ringpop TChannel", tag.Error(err))
+						} else {
+							if err := tChannel.Serve(listener); err != nil {
+								factory.logger.Fatal("Failed to serve ringpop listener", tag.Error(err), tag.Address(ringpopHostAddress))
+							}
+							factory.channel = tChannel
+						}
+					}
+				}
+			}
+		} else {
+			if listener, err := net.Listen("tcp", ringpopHostAddress); err != nil {
+				factory.logger.Fatal("Failed to start ringpop listener", tag.Error(err), tag.Address(ringpopHostAddress))
+			} else {
+				if tChannel, err := tchannel.NewChannel(ringpopServiceName, &tchannel.ChannelOptions{}); err != nil {
+					factory.logger.Fatal("Failed to create ringpop TChannel", tag.Error(err))
+				} else {
+					if err := tChannel.Serve(listener); err != nil {
+						factory.logger.Fatal("Failed to serve ringpop listener", tag.Error(err), tag.Address(ringpopHostAddress))
+					}
+					factory.channel = tChannel
+				}
+			}
+		}
+	})
+
+	return factory.channel
 }
 
-func (factory *RingpopFactory) getRingpop() (*membership.RingPop, error) {
-	if factory.ringPop != nil {
-		return factory.ringPop, nil
+func (factory *ringpopFactory) getListenIP() net.IP {
+	if factory.rpcConfig.BindOnLocalHost && len(factory.rpcConfig.BindOnIP) > 0 {
+		factory.logger.Fatal("ListenIP failed, bindOnLocalHost and bindOnIP are mutually exclusive")
+		return nil
 	}
 
-	ringPop, err := factory.createRingpop()
-	if err != nil {
-		return nil, err
+	if factory.rpcConfig.BindOnLocalHost {
+		return IPV4Localhost
 	}
-	factory.ringPop = ringPop
-	return ringPop, nil
-}
 
-func (factory *RingpopFactory) broadcastAddressResolver() (string, error) {
-	return membership.BuildBroadcastHostPort(factory.channel.PeerInfo(), factory.config.BroadcastAddress)
-}
-
-func (factory *RingpopFactory) createRingpop() (*membership.RingPop, error) {
-	rp, err := ringpop.New("temporal", ringpop.Channel(factory.channel), ringpop.AddressResolverFunc(factory.broadcastAddressResolver))
+	if len(factory.rpcConfig.BindOnIP) > 0 {
+		ip := net.ParseIP(factory.rpcConfig.BindOnIP)
+		if ip != nil {
+			return ip
+		}
+		factory.logger.Fatal("ListenIP failed, unable to parse bindOnIP value", tag.Address(factory.rpcConfig.BindOnIP))
+		return nil
+	}
+	ip, err := config.ListenIP()
 	if err != nil {
-		return nil, err
+		factory.logger.Fatal("ListenIP failed", tag.Error(err))
+		return nil
 	}
+	return ip
+}
 
-	return membership.NewRingPop(rp, factory.config.MaxJoinDuration, factory.logger), nil
+// CloseTChannel allows fx Stop hook to close channel
+func (factory *ringpopFactory) CloseTChannel() {
+	if factory.channel != nil {
+		factory.getTChannel().Close()
```

Comment:
should you set factory.channel = nil after close?

Reply:
Probably yes to be safe, but this is only called during fx Lifecycle hooks.